server/src/main/java/org/cloudfoundry/identity/uaa/client/ClientJwtConfiguration.java

```diff
@@ -0,0 +1,308 @@
+    return this.jwkSet;
+  }
+
+  public void setJwkSet(final JsonWebKeySet<JsonWebKey> jwkSet) {
+    this.jwkSet = jwkSet;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
+
+    if (o instanceof ClientJwtConfiguration) {
+      ClientJwtConfiguration that = (ClientJwtConfiguration) o;
+      if (!Objects.equals(jwksUri, that.jwksUri)) return false;
+      if (jwkSet != null && that.jwkSet != null) {
+        return jwkSet.getKeys().equals(that.jwkSet.getKeys());
+      } else {
+        return Objects.equals(jwkSet, that.jwkSet);
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public int hashCode() {
+    int result = super.hashCode();
+
+    result = 31 * result + (jwksUri != null ? jwksUri.hashCode() : 0);
+    result = 31 * result + (jwkSet != null ? jwkSet.hashCode() : 0);
+    return result;
+  }
+
+  @Override
+  public Object clone() throws CloneNotSupportedException {
+    return super.clone();
+  }
+
+  @JsonIgnore
+  public String getCleanString() {
+    try {
+      if (UaaUrlUtils.isUrl(this.jwksUri)) {
+        return this.jwksUri;
+      } else if (this.jwkSet != null && !ObjectUtils.isEmpty(this.jwkSet.getKeySetMap())) {
+        return JWKSet.parse(this.jwkSet.getKeySetMap()).toString(true);
+      }
+    } catch (IllegalStateException | JsonUtils.JsonUtilException | ParseException e) {
+      throw new InvalidClientDetailsException("Client jwt configuration configuration fails ", e);
+    }
+    return null;
+  }
+
+  @JsonIgnore
+  public static ClientJwtConfiguration parse(String privateKeyConfig) {
+    if (UaaUrlUtils.isUrl(privateKeyConfig)) {
+      return parse(privateKeyConfig, null);
+    } else {
+      return parse(null, privateKeyConfig);
+    }
+  }
+
+  @JsonIgnore
+  public static ClientJwtConfiguration parse(String privateKeyUrl, String privateKeyJwt) {
+    ClientJwtConfiguration clientJwtConfiguration = null;
+    if (privateKeyUrl != null) {
+      clientJwtConfiguration = new ClientJwtConfiguration(privateKeyUrl, null);
+      clientJwtConfiguration.validateJwksUri();
+    } else if (privateKeyJwt != null && privateKeyJwt.contains("{") && privateKeyJwt.contains("}")) {
+      HashMap<String, Object> jsonMap = JsonUtils.readValue(privateKeyJwt, HashMap.class);
+      String cleanJwtString;
+      try {
+        if (jsonMap.containsKey("keys")) {
+          cleanJwtString = JWKSet.parse(jsonMap).toString(true);
+        } else {
+          cleanJwtString = JWK.parse(jsonMap).toPublicJWK().toString();
+        }
+        clientJwtConfiguration = new ClientJwtConfiguration(null, JsonWebKeyHelper.deserialize(cleanJwtString));
+        clientJwtConfiguration.validateJwkSet();
+      } catch (ParseException e) {
+        throw new InvalidClientDetailsException("Client jwt configuration cannot be parsed", e);
+      }
+    }
+    return clientJwtConfiguration;
+  }
```

Comment:
It would make more sense to split this method to two separate ones, one for parsing `privateKeyUrl` and the other for parsing `privateKeyJwt`. I don't see any code shared between the two.

Reply:
allowed is one, either URI or JWK itself, therefore I wanted to have it in one class.

Reply:
The question was not about the class code but about having one method doing two different things. Not a big enough problem to block this PR, though. So I am find with not changing the method.

Reply:
i did some refactoring, if you have improvements you can do pushes